### ISSUE TITLE
rsyslogd version 8.23 leaks memory when TCP connections are reset

### ISSUE THREAD

**Issue by tylerschultz** [rsyslog/rsyslog#1334]
More details can be found [here](https://github.com/cloudfoundry/bosh/issues/1537)

**Comment by holgero** [comment]
I tried to reproduce something manually (by letting logger write some logs, configured rsyslog to forward to 127.0.0.1:5555 and starting and killing netcat several times). Here is the output of valgrind during that experiments. It seems to report something, but I don't know if that means anything (I am not familiar with valgrind (yet)). Anyway, here is what I got: [valgrind.txt](https://github.com/rsyslog/rsyslog/files/724429/valgrind.txt)

**Comment by voelzmo** [comment]
cc @rgerhards

**Comment by holgero** [comment]
I have an more advanced setup now to reproduce the problem:
* rsyslog is configured to forward to 127.0.0.1:5555.
* haproxy listens on that port and cannot forward it anywhere (so it closes the connection again immediately, `netstat -tupen` shows a lot of these in TIME_WAIT)
* logger logs in a shell loop a random string (with a delay of .5 seconds)

I could see with this setup the amount of memory used by the rsyslog process increasing slowly, here are my notes of that session (without valgrind):
````
  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND                                                                

Tue Jan 24 12:35:53 UTC 2017
 6424 syslog    10 -10  493948   4728   3396 S   0.0  0.0   0:00.48 rsyslogd                                                               
Tue Jan 24 12:36:20 UTC 2017
 6424 syslog    10 -10  493948   4744   3396 S   0.0  0.0   0:00.51 rsyslogd                                                               
Tue Jan 24 12:37:26 UTC 2017
 6424 syslog    10 -10  493948   4936   3396 S   0.3  0.0   0:00.65 rsyslogd                                                               
Tue Jan 24 12:40:04 UTC 2017
 6424 syslog    10 -10  493948   5324   3396 S   0.0  0.1   0:00.95 rsyslogd                                                               
Tue Jan 24 12:40:53 UTC 2017
 6424 syslog    10 -10  493948   5356   3396 S   0.0  0.1   0:00.99 rsyslogd                                                               

shutting down loggger
Tue Jan 24 12:41:52 UTC 2017
 6424 syslog    10 -10  493948   5400   3396 S   0.0  0.1   0:01.03 rsyslogd                                                               
Tue Jan 24 12:43:33 UTC 2017
 6424 syslog    10 -10  493948   5404   3396 S   0.0  0.1   0:01.04 rsyslogd                                                               
Tue Jan 24 12:49:04 UTC 2017
 6424 syslog    10 -10  493948   5404   3396 S   0.0  0.1   0:01.04 rsyslogd                                                               

starting logger again
Tue Jan 24 12:49:33 UTC 2017
 6424 syslog    10 -10  493948   5416   3396 S   0.0  0.1   0:01.05 rsyslogd                                                               
Tue Jan 24 12:50:25 UTC 2017
 6424 syslog    10 -10  493948   5512   3396 S   0.3  0.1   0:01.14 rsyslogd                                                               
Tue Jan 24 12:51:16 UTC 2017
 6424 syslog    10 -10  493948   5552   3396 S   0.0  0.1   0:01.18 rsyslogd                                                               
Tue Jan 24 12:52:00 UTC 2017
 6424 syslog    10 -10  493948   5656   3396 S   0.0  0.1   0:01.26 rsyslogd                                                               
Tue Jan 24 12:55:25 UTC 2017
 6424 syslog    10 -10  493948   6036   3396 S   0.7  0.1   0:01.60 rsyslogd                                                               
Tue Jan 24 12:58:06 UTC 2017
 6424 syslog    10 -10  493948   6516   3396 S   0.7  0.1   0:01.96 rsyslogd                                                               
Tue Jan 24 12:58:33 UTC 2017
 6424 syslog    10 -10  493948   6532   3396 S   0.0  0.1   0:01.97 rsyslogd                                                               

stopped logger again
Tue Jan 24 12:58:47 UTC 2017
 6424 syslog    10 -10  493948   6532   3396 S   0.0  0.1   0:01.97 rsyslogd                                                               
````

One can see how RES increases slowly while the logger shell loop runs.

Afterwards I ran the setup with valgrind, here is the logfile: [valgrind2.txt](https://github.com/rsyslog/rsyslog/files/726785/valgrind2.txt)

**Comment by rgerhards** [comment]
Thanks, this experiment seems to be very useful. Unfortunately, the otherwise very useful valgrind log does not contain most symbol information. This is due to the unloading of modules at end of run.

It would be great if you could add

```global(debug.unloadModules="off")```

to your rsyslog.conf and re-run the experiment.

**Comment by holgero** [comment]
sure, I'll do this

**Comment by holgero** [comment]
Hmm, didn't seem to have any effect. The rest of rsyslog.conf looks quite different syntaxwise from what you suggested, though. All things start with $-signs, can I simply add the line you gave me there?

**Comment by holgero** [comment]
I just had a look at the executable, and it is stripped:
````
/:/etc/rsyslog.d# file /usr/sbin/rsyslogd 
/usr/sbin/rsyslogd: ELF 64-bit LSB  executable, x86-64, version 1 (SYSV), dynamically linked (uses shared libs), for GNU/Linux 2.6.24, BuildID[sha1]=e52673c29cc8284ce49953089a7bde932d934d06, stripped
````

**Comment by rgerhards** [comment]
> All things start with $-signs, can I simply add the line you gave me there?

That's an old-style config, unfortunately still the default with many distros. But new and old style can be mixed.

Can you build rsyslog yourself? If not, I'll need to see that we reproduce the issue here - or find another debug means (maybe a debug log would help, let me think about it).

**Comment by holgero** [comment]
Never built it myself. If it's not too complicated I could give it a try.

**Comment by holgero** [comment]
configure: error: Package requirements (libfastjson >= 0.99.3) were not met:
No package 'libfastjson' found

I didn't find a package with libfastjson on ubuntu xenial, so I'd rather not continue to try building it myself. Any other options?

**Comment by rgerhards** [comment]
we do have these things in the Adiscon/rsyslog PPA, but let me try first if I can reproduce the issue with some new testbench tooling. That would have the advantage that we can include it in CI testing. I guess I can craft something based on your description.

**Comment by holgero** [comment]
:+1: :smiley:

**Comment by rgerhards** [comment]
mhhh... I wrote a test program that drops n connections, and only then accepts data. Unfortunately, this does not lead to a memory leak like you experience. I guess I need to  re-model the way I test. Maybe we actually need a debug log from your env... I keep you posted.

**Comment by Whissi** [comment]
What's the current status of this issue? I set stabilization of >rsyslog-8.19.0 on hold due to this report.
Was this issue ever confirmed?

**Comment by rgerhards** [comment]
I could unfortunately but reproduce it, even though I created some very
specific tests.

Rainer

Sent from phone, thus brief.

Am 14.02.2017 15:51 schrieb "Thomas Deutschmann" <notifications@github.com>:

**Comment by voelzmo** [comment]
@rgerhards so you're saying the setup described  in https://github.com/rsyslog/rsyslog/issues/1334#issuecomment-274802288 didn't expose the issues for you?

**Comment by rgerhards** [comment]
I don't have haproxy, but I wrote a program that creates the same condition. There I could not see the issue. I don't have time right now to learn how to setup haproxy just to do that test, sorry.

**Comment by voelzmo** [comment]
here is an haproxy cfg for you
```
global
#    log 127.0.0.1   syslog info
    daemon
    user vcap
    group vcap
    maxconn 64000
    spread-checks 4
    tune.bufsize 16384


defaults
    log global
    timeout connect 30000ms
    maxconn 64000

    timeout client 900000ms
    timeout server 900000ms




frontend http-in
    mode http
    bind :80
    option httplog
    option forwardfor
    reqadd X-Forwarded-Proto:\ http
    default_backend http-routers


listen health_check_http_url
    bind :8080
    mode http
    monitor-uri /health


frontend log-in
    mode tcp
    bind :5555
    default_backend tcp-routers

backend http-routers
    mode http
    balance roundrobin



backend tcp-routers
    mode tcp
    balance roundrobin
```

**Comment by anEXPer** [comment]
For what it's worth, we (folks working on Cloud Foundry) recently came to the conclusion that the HA Proxy-based replication was invalid anyway.

The issue is that the "slow increase in allocated memory" reported here isn't a memory leak, because it:
- has bounded growth
- is released according to business/configuration rules

It's just the action queue growing to its configured size as a linked list in the absence of any valid connection over which to discharge its accumulating actions.

HA Proxy without a functioning backend accepts and then quickly kills TCP connections, never  accepting data. We ran the replication case for longer than ten minutes, and the memory usage stabilized and held steady at around 1.4% of available memory.

If all the replication shows is the growth of action queues, we'd expect the memory reservation to drop as soon as a single good connection was made - and when we performed that experiment (by reconfiguring the HA Proxy to forward to an rsyslog configured to receive and store logs), that's exactly what we saw - a large batch of logs hit the rsyslog configured to receive, and memory usage dropped to nominal levels on the vm with an rsyslog configured to forward.
